### DISCUSSION TITLE
Question about history handling

### DISCUSSION THREAD

**Discussion by jkemp814** [ohmybash/oh-my-bash#500] [Q&A] [answered]
Hello Everyone,

Great piece of software.

I have a question about the way history is being handled. Coming from plain bash, I had many lines pertaining to history. When installing OMB I renamed .bashrc to .bashrc.orig. I know it would have backed up the original file, but I renamed it anyway.

I am still getting duplicates in my .bash_history file. Does the /lib/history.sh library handle all the duplicates or do I still need to add something to the .bashrc file? On line 35 and 36 of the /lib/history.sh file, there is no 'export' to HISTCONTROL=. Maybe it doesn't need it, I'm no expert. Does it only handle the history in memory? I typed 'git status' three times and notice it did not add all three. But if I return and open a new terminal window, it will add another one and not remove the existing one. So far, I have many of them in the .bash_history file. How do I get rid of those. Like I mention, I did not add much to the bashrc file. Just some cargo/env and fzf stuff. Nothing pertaining to history.

**Comment by akinomyoga** [comment, marked as answer]
> I am still getting duplicates in my .bash_history file.

This is a consequence of `history -a` called in PROMPT_COMMAND and in theme definitions. The new entry is written to the history file immediately after the command is executed. So, even if the duplicate entries are removed from the in-memory history by `erasedups`, the entries that are already written to files remain. Since `shopt -s histappend` is specified, the history file will not be retroactively rewritten on the session end.

To force the `erasedups` to the history file, you need to turn off the synchronized history; You need to comment out line 27 of `lib/history.sh`, and also remove `history -a;history -c;history -r` in each theme file if any. Or, if you would like to keep the history synchronization, another way would be to manually remove the duplicate entries of the history file when you want it.

> On line 35 and 36 of the /lib/history.sh file, there is no 'export' to HISTCONTROL=. Maybe it doesn't need it, I'm no expert.

It's unrelated. There is no need to export `HISTCONTROL`. `HISTCONTROL` shouldn't be exported unless you know what you want to achieve by exporting `HISTCONTROL`. In this sense, I think `HISTIGNORE` shouldn't be exported either although it is now exported on line 39.

>  So far, I have many of them in the .bash_history file. How do I get rid of those.

See https://unix.stackexchange.com/a/78846/121088. In case the timestamp is also included in the history file, you would probably need to preserve the lines starting with `# ` as well.

```bash
$ cp ~/.bash_history ~/.bash_history.backup
$ tac ~/.bash_history.backup | awk '/^# / || !mark[$0]++' | tac > ~/.bash_history
```

If you are in Solaris, please use `/usr/xpg4/bin/awk` instead of `awk` since the default awk in Solaris is not POSIX awk and doesn't have sufficient features.

**Comment by jkemp814** [comment]
Great answer. You really summed it up. Thank you.

I have also been using ble.sh. The auto-complete and suggestions are working great. It even suggested commands with no history at all. I use a powerful little log navigator called LNAV as soon as I started to type `journalctl -f | lnav` it suggested `journalctl`. 

I noticed ble.sh has some lines of code related to history also. In your opinion should I be using both? I haven't uncommented any of the settings in ble.sh in fear it may create some race condition. Are there plans to merge all these technologies?

All of your hard work on these Bash projects is really appreciated.

**Comment by jkemp814** [comment]
Check out this little gem here. Very tiny but quite effective (except the ghosted timestamps)

https://gitlab.com/vn971/shell-history-cleaner
